tests/integ/ml/test_transformer.py

```diff
@@ -0,0 +1,346 @@
+            Row(
+                A=2,
+                B=3,
+                C=6,
+                ONEHOTARRAY="[\n  6,\n  [\n    1,\n    3,\n    5\n  ],\n  1\n]",
+            ),
+        ],
+    )
+
+
+def test_sparse_onehotencoder(session):
+    df = session.createDataFrame(
+        [[1, 4, 5], [2, 4, 5], [2, 3, 6]], schema=["a", "b", "c"]
+    )
+    onehotencoder = OneHotEncoder()
+    onehotencoder.input_cols = ["A", "B", "C"]
+    onehotencoder.output_cols = ["A_ONEHOT", "B_ONEHOT", "C_ONEHOT"]
+    model = onehotencoder.fit(df)
+
+    # test category table
+    df_table = session.table(model._category_table_name["A"])
+    Utils.check_answer(
+        df_table,
+        [Row(STATES_A=1, STATES_A_ENCODER=0), Row(STATES_A=2, STATES_A_ENCODER=1)],
+    )
+
+    res = model.transform(df, sparse=True)
+    Utils.check_answer(
+        res,
+        [
+            Row(
+                A=1,
+                B=4,
+                C=5,
+                A_ONEHOT="[\n  1,\n  0\n]",
+                B_ONEHOT="[\n  1,\n  0\n]",
+                C_ONEHOT="[\n  1,\n  0\n]",
+            ),
+            Row(
+                A=2,
+                B=3,
+                C=6,
+                A_ONEHOT="[\n  0,\n  1\n]",
+                B_ONEHOT="[\n  0,\n  1\n]",
+                C_ONEHOT="[\n  0,\n  1\n]",
+            ),
+            Row(
+                A=2,
+                B=4,
+                C=5,
+                A_ONEHOT="[\n  0,\n  1\n]",
+                B_ONEHOT="[\n  1,\n  0\n]",
+                C_ONEHOT="[\n  1,\n  0\n]",
+            ),
+        ],
+    )
+
+
+def test_sparse_merge_onehotencoder(session):
+    df = session.createDataFrame(
+        [[1, 4, 5], [2, 4, 5], [2, 3, 6]], schema=["a", "b", "c"]
+    )
+    onehotencoder = OneHotEncoder()
+    onehotencoder.input_cols = ["A", "B", "C"]
+    onehotencoder.output_cols = ["ONEHOTARRAY"]
+    model = onehotencoder.fit(df)
+    # test category table
+    df_table = session.table(model._category_table_name["A"])
+    Utils.check_answer(
+        df_table,
+        [Row(STATES_A=1, STATES_A_ENCODER=0), Row(STATES_A=2, STATES_A_ENCODER=1)],
+    )
+
+    res = model.transform(df, sparse=True)
+    Utils.check_answer(
+        res,
+        [
+            Row(A=1, B=4, C=5, ONEHOTARRAY="[\n  1,\n  0,\n  1,\n  0,\n  1,\n  0\n]"),
+            Row(A=2, B=3, C=6, ONEHOTARRAY="[\n  0,\n  1,\n  0,\n  1,\n  0,\n  1\n]"),
+            Row(A=2, B=4, C=5, ONEHOTARRAY="[\n  0,\n  1,\n  1,\n  0,\n  1,\n  0\n]"),
+        ],
+    )
+
+
+def test_input_cols_output_cols_not_match(session):
+    df = session.createDataFrame(
+        [[1, 4, 5], [2, 4, 5], [2, 3, 6]], schema=["a", "b", "c"]
+    )
+    onehotencoder = OneHotEncoder()
+    onehotencoder.input_cols = ["A", "B", "C"]
+    onehotencoder.output_cols = ["B", "C"]
+    model = onehotencoder.fit(df)
+    with pytest.raises(ValueError):
+        model.transform(df)
+
+
+def test_preset_category_not_match_real_category(session):
```

Comment:
This edge case doesn't only happen when we have preset category, right? For example, 
```
df = session.createDataFrame(
        [[1, 4, 5], [2, 4, 5], [2, 3, 6]], schema=["a", "b", "c"]
    )
```
If we get 7 in transform() call, it should be handled correctly. Also, let's check this edge case for 
1. onehot encoder 
   1.a sparse
   1.b not sparse 
   (you can also test merge, but not very necessary)
   1.c preset category
   1.d auto mode
2. ordinal encoder
   1.a preset category
   1.b auto mode